app/src/main/java/org/kiwix/kiwixmobile/webserver/ZimHostActivity.java

```diff
@@ -0,0 +1,544 @@
+    super.onResume();
+    presenter.loadBooks();
+    if (isServerStarted) {
+      ip = getAddress();
+      ip = ip.replaceAll("\n", "");
+      serverTextView.setText(getString(R.string.server_started_message, ip));
+      startServerButton.setText(getString(R.string.stop_server_label));
+      startServerButton.setBackgroundColor(getResources().getColor(R.color.stopServer));
+    }
+  }
+
+  // This method checks if mobile data is enabled in user's device.
+  static boolean isMobileDataEnabled(Context context) {
+    ConnectivityManager cm =
+        (ConnectivityManager) context.getSystemService(Context.CONNECTIVITY_SERVICE);
+    try {
+      Class cmClass = Class.forName(cm.getClass().getName());
+      Method method = cmClass.getDeclaredMethod("getMobileDataEnabled");
+      method.setAccessible(true);
+      return (Boolean) method.invoke(cm);
+    } catch (Exception e) {
+      Log.e(TAG, e.toString());
+    }
+    return false;
+  }
+
+  //This method sends the user to data usage summary settings activity
+  private void openMobileDataActivity() {
+    startActivity(new Intent(android.provider.Settings.ACTION_WIRELESS_SETTINGS));
+  }
+
+  @Override public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions,
+      @NonNull int[] grantResults) {
+    if (requestCode == MY_PERMISSIONS_ACCESS_FINE_LOCATION) {
+      if (grantResults.length > 0
+          && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+          toggleHotspot();
+        }
+      }
+    }
+  }
+
+  @Override
+  protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
+    //Checking the result code for LocationSettings resolution
+    if (requestCode == LOCATION_SETTINGS_PERMISSION_RESULT) {
+      final LocationSettingsStates states = LocationSettingsStates.fromIntent(data);
+      switch (resultCode) {
+        case Activity.RESULT_OK:
+          // All required changes were successfully made
+          Log.v(TAG, states.isLocationPresent() + "");
+          startService(ACTION_TURN_ON_AFTER_O);
+          break;
+        case Activity.RESULT_CANCELED:
+          // The user was asked to change settings, but chose not to
+          Log.v(TAG, "Canceled");
+          break;
+        default:
+          break;
+      }
+    }
+  }
+
+  @Override protected void onDestroy() {
+    super.onDestroy();
+    presenter.detachView();
+  }
+
+  private void setUpToolbar() {
+    Toolbar toolbar = findViewById(R.id.toolbar);
+    setSupportActionBar(toolbar);
+    getSupportActionBar().setTitle(getString(R.string.menu_host_books));
+    getSupportActionBar().setHomeButtonEnabled(true);
+    getSupportActionBar().setDisplayHomeAsUpEnabled(true);
+
+    toolbar.setNavigationOnClickListener(v -> onBackPressed());
+  }
+
+  private void setupLocationServices() {
+    LocationRequest locationRequest = new LocationRequest();
```

Comment:
I would like to see all this location stuff handled in another class

Reply:
But we need the response in an `Activity`. i.e. `startResolutionForResult`. To compare if everything is evaluated correctly, and call appropriate methods in the `ZimHostActivity`.

Reply:
Define callbacks, forward events from ZimHostActivity to the new class.

Reply:
Fixed. 62b7a99e18b09d6d73b633c7f1eea4171913e9f8